py_pdf_parser/filtering.py

```diff
@@ -213,7 +213,23 @@ def filter_by_fonts(self, *fonts: str) -> "ElementList":
         Returns:
             ElementList: The filtered list.
         """
-        new_indexes = set(element._index for element in self if element.font in fonts)
+        non_cached_fonts = fonts - self.document._elements_indexes_by_font.keys()
+        if non_cached_fonts:
+            # If we don't have cached elements for any of the required fonts, build
+            # the cache for the non cached fonts.
+            for element in self:
+                if element.font not in non_cached_fonts:
+                    continue
+
+                self.document._elements_indexes_by_font[element.font].add(
+                    element._index
+                )
```

Comment:
I'm afraid I don't think this works.

Say I have element A on page 1 and element B on page 2, both with the same font.

If I do `document.get_page(1).elements.filter_by_font(font)`, then element A gets added to the cache.
Then when I do `document.get_page(2).elements.filter_by_font(font)`, `non_cached_fonts` will be empty, and you code will return an empty element list since B isn't in the cache?

I think you want something like:
```
        non_cached_fonts = fonts - self.document._elements_indexes_by_font.keys()
        if non_cached_fonts:
            # If we don't have cached elements for any of the required fonts, build
            # the cache
            self.document._elements_indexes_by_font[font] = set(element._index for element in document.elements if element.font == font)
```

However, I also think this might be cleaner as a private method on the document? So e.g. `document._elements_with_font(font)`, which uses the `__element_indexes_by_font` variable? It seems weird to me that a function external to the `PDFDocument` is handling a cache on the `PDFDocument`, if that makes sense?

Reply:
Well spotted, I forgot that I was actually looping the elements list and not the document itself!